fri/src/matrix_reducer.rs

```diff
@@ -0,0 +1,188 @@
+        Self {
+            alpha,
+            alpha_pow_width: alpha.exp_u64(F::Packing::WIDTH as u64),
+            transposed_alphas,
+            current_alpha_pow: EF::one(),
+        }
+    }
+
+    #[instrument(name = "fold in matrices", level = "debug", skip(self, reduced, mats))]
+    pub(crate) fn reduce_matrices<M: MatrixRows<F> + Sync>(
+        &mut self,
+        reduced: &mut [EF],
+        height: usize,
+        mats: &[M],
+    ) {
+        // precompute alpha_pows, since we are not doing horner
+        let mut alpha_pows = vec![];
+        info_span!("precompute alphas").in_scope(|| {
+            for mat in mats {
+                let num_packed = mat.width() / F::Packing::WIDTH;
+                let num_leftover = mat.width() % F::Packing::WIDTH;
+                for chunk in &(0..num_packed).chunks(BATCH_SIZE) {
+                    alpha_pows.push(self.current_alpha_pow);
+                    self.current_alpha_pow *= self.alpha_pow_width.exp_u64(chunk.count() as u64);
+                }
+                for _ in 0..num_leftover {
+                    alpha_pows.push(self.current_alpha_pow);
+                    self.current_alpha_pow *= self.alpha;
+                }
+            }
+        });
+
+        /*
+        We have packed base elements ys and extension field α^n's
+        We want α^0 * ys[0] + α^1 * ys[1] + ...
+
+        Level 0:
+            alpha.powers().zip(cols).map(|alpha_pow, col| alpha_pow * col).sum()
+
+        Level 1: (assume Packing::WIDTH=4, D=2, for clarity, although D will usually be higher)
+
+            tranposed_alphas
+            α^0 α^1 α^2 α^3
+            [1] [.] [.] [.]
+            [1] [.] [.] [.]
+         ys:[a   b   c   d] [e f g h] ...
+
+        We multiply ys vertically, then sum horizontally.
+        Aka, if β is α^0*a + α^1*b + α^2*c + α^3*d, then each limb of β is
+        determined by multiplying the packed ys by the appropriate row of
+        transposed_alphas, then summing horizontally.
+
+        This assumes we are in an extension field where multiplication and addition
+        by a base element is simply elementwise.
+
+        Then, to fold it into our running reduction, we perform one extension multiplication.
+        In this scheme, the extension mul takes about 40% of the time in this function.
+
+        Level 2: Batching. To delay the horizontal sum and extension mul as much as possible,
+        we precompute even more columns of transposed_alphas, and sum the packed results per
+        batch, and only do one horizontal sum and extension mul per batch.
+
+        */
+
+        info_span!("reduce").in_scope(|| {
+            reduced
+                .into_par_iter()
+                // .into_iter()
+                .enumerate()
+                .for_each(|(r, reduced_row)| {
+                    let mut alpha_pow_iter = alpha_pows.iter();
+                    for mat in mats {
+                        let row_slice = mat.row(r).into_iter().collect_vec();
```

Comment:
going to add a method on `MatrixRows` that returns a `Cow<&[F]>`

Reply:
well the alternative is some bespoke `PackedIterator` that also returns leftovers

Reply:
nvm just added `MatrixRows::row_vec` that allocates if not overridden